File: primes/src/primes_omp.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <omp.h>

int main(int argc, char *argv[])
{
  unsigned long long int i,j,limit,Nprimes;
  bool *primes;
  double start_t,end_t;

  limit=500000000;
  Nprimes=0;
  start_t=omp_get_wtime();

  primes = malloc(sizeof(bool)*limit);
  primes[0]=0;
  primes[1]=0;
#pragma omp parallel default(none) shared(primes,Nprimes) private(j) firstprivate(limit)
  {
    /* fill boolean array */
#pragma omp for schedule(static)
    for (i=2;i<limit;i++)
      {
	primes[i]=1;
      }
    /* Sieve of Eratosthenes */
#pragma omp for schedule(static)
    for (i=2;i<limit;i++)
      {
	if (primes[i])
	  {
	    for (j=i;i*j<limit;j++)
	      {
		primes[i*j]=0;
	      }
	  }
      }
#pragma omp for reduction(+:Nprimes) schedule(static)
    for(i=2;i<limit;i++)
      {
	if (primes[i]) Nprimes++;
      }
  }
  
  end_t=omp_get_wtime();
  printf("Nprimes = %llu\n",Nprimes);
  printf("Took %g s\n",end_t - start_t);
  return EXIT_SUCCESS;
}

File: primes/src/primes_omp.f90
PROGRAM primes_omp
  USE omp_lib
  IMPLICIT NONE
  INTEGER(kind=8) ::  i, Nprimes
  REAL(kind=8) :: start_t, end_t
  LOGICAL, ALLOCATABLE :: primes(:)
  INTEGER, PARAMETER :: limit = 500000000

  start_t = omp_get_wtime()

  ! Initialise 
  ALLOCATE(primes(limit))
  primes=.true.
  primes(1)=.false.
  
  ! Sieve of Eratosthenes
  
  !$OMP PARALLEL DEFAULT(none)    &
  !$OMP SHARED(primes,Nprimes)
  !$OMP DO SCHEDULE(static)
  DO i=2,limit
     IF (primes(i)) THEN
        primes(i*i : limit : i) = .false.
     END IF
  END DO
  !$OMP END DO

  ! Count up number of primes
  Nprimes=0
  !$OMP DO SCHEDULE(static) REDUCTION(+:Nprimes)
  DO i=1,limit
     IF (primes(i)) Nprimes = Nprimes+1
  END DO
  !$OMP END DO
  !$OMP END PARALLEL
  
  end_t = omp_get_wtime()

  WRITE(*,*) "Number of primes = ",Nprimes
  WRITE(*,*) "Expended wall clock time = ",end_t-start_t  
  
END PROGRAM primes_omp
